feat(ecmascript): proxy internal get prototype of

yossydev opened the pull request:
I have implemented the internal_get_prototype_of method for the Proxy object based on the ECMAScript specification: https://tc39.es/ecma262/#sec-proxy-object-internal-methods-and-internal-slots-getprototypeof.

However, to pass the test262 tests, it seems necessary to enable the creation of Proxy instances. As a result, the implementation is currently incomplete, focusing only on the internal logic of the method for now.

You can find the relevant implementation here: https://github.com/trynova/nova/blob/5a698d79c26f92890b3ed355a317c3c6a7e78b05/nova_vm/src/ecmascript/builtins/reflection/proxy_constructor.rs#L38-L46

ref: https://github.com/trynova/nova/issues/160

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/data.rs:
```diff
@@ -3,13 +3,15 @@
 
 #[derive(Debug, Clone)]
 pub struct ProxyHeapData {
     pub(crate) object_index: Option<OrdinaryObject<'static>>,
```
**issue:** Since we're here, let's remove the object_index from ProxyHeapData. It will never be used and thus makes no sense here.

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/data.rs:
```diff
@@ -3,13 +3,15 @@
 #[derive(Debug, Clone)]
 pub struct ProxyHeapData {
     pub(crate) object_index: Option<OrdinaryObject<'static>>,
+    pub(crate) target: Option<Value>,
```
**nitpick:** Let's keep strictly with the spec and name these `proxy_target` and `proxy_handler` so as to reflect `[[ProxyTarget]` and `[[ProxyHandler]]]`.

**suggestion:** Adding doc comments (`///`)  to also note that these are the `[[ProxyTaget]]` and `[[ProxyHandler]]` respectively would also be a nice touch.

yossydev replied:
Thank you for the review! You’re absolutely right, that’s how it should have been! I’ve fixed it accordingly! https://github.com/trynova/nova/pull/494/commits/ce103d3cf5533fb1b4d1b52ca0794f16481e0237

yossydev replied:
**question**: By the way, this part uses target and handler, but is that okay? Apologies if I’m saying something strange.

https://github.com/trynova/nova/blob/c0feafd4d49916ec8ac0bac5441e3575a2a8737e/nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs#L15-L19

aapoalas replied:
Yeah, I pondered the same thing myself but figured that it doesn't matter too much, and since the return values are always assigned to names "target" and "handler", it's nice to be able to directly destructure the struct.

Effectively, either is fine so let's leave it how it happens to be now :)

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/data.rs:
```diff
@@ -3,13 +3,15 @@
 #[derive(Debug, Clone)]
 pub struct ProxyHeapData {
     pub(crate) object_index: Option<OrdinaryObject<'static>>,
+    pub(crate) target: Option<Value>,
```
**issue:** According to spec, `[[ProxyTarget]]` is "either an object or the null value". In Nova that would be `Option<Object>`.

Similarly, `[[ProxyHandler]]` is also said to be "an object, called the proxy's handler object, or null", so it should also be `Option<Object>`.

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+        }
+
+        // 7. Let handlerProto be ? Call(trap, handler, « target »).
+        let handler_proto = call(
```
**issue:** It's better to use `call_function` here, as `get_method` returns us a `Function` that we can call without performing some of the checks that `call` needs to do.

yossydev replied:
This hadn’t been fixed yet, so I went ahead and corrected it!
https://github.com/trynova/nova/pull/494/commits/aa4cca1d0dd52f0cb4047104b5e02627f51f64e7

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs:
```diff
@@ -0,0 +1,40 @@
+/// The abstract operation ValidateNonRevokedProxy takes argument
+/// proxy (a Proxy exotic object) and returns either a normal completion containing unused or a throw completion.
+/// It throws a TypeError exception if proxy has been revoked.
+pub(crate) fn validate_non_revoked_proxy(
```
**suggestion:** It would be useful here to actually return both `target` and `handler` from the method. From the spec text point of view that's a little "Invalid" but from a "spirit of the spec" standpoint it's absolutely the correct thing to do: This method validates that both the handler and target are non-null, so it then makes sense to return those values so that callers don't need to re-fetch and `unwrap()` them all over again.

yossydev replied:
I see, I wrote it to align with the specification for now, but it’s good to know that’s also an option. That’s very helpful, thank you!

aapoalas replied:
Yeah, basically whenever the spec uses an "assert function" ie. a function that checks something about it's parameters and either returns UNUSED or an error, then we'll often want to change that to returning the "asserted" data.

As an example, IsCallable currently returns a Function because that's what the real intention of the method is (except that once callable Proxies are a thing, we'll need a new Callable type...). IsObject probably should return an object etc.

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs:
```diff
@@ -0,0 +1,40 @@
+    let proxy_data = &agent[proxy];
+
+    // 1. If proxy.[[ProxyTarget]] is null, throw a TypeError exception.
+    if proxy_data.target.is_none() {
```
```suggestion
    let Some(target) = proxy_data.target else {
```

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    // 2. Assert: proxy.[[ProxyHandler]] is not null.
+    if proxy_data.handler.is_none() {
```
```suggestion
    let Some(handler) = proxy_data.handler else {
```

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    // 3. Return unused.
+    Ok(())
```
```suggestion
    Ok(NonRevokedProxy { handler, target })
```

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+
+        // 9. Let extensibleTarget be ? IsExtensible(target).
+        let extensible_target = if let ControlFlow::Continue(result) =
+            try_is_extensible(agent, Object::try_from(target.unwrap()).unwrap(), gc.nogc())
```
**issue:** You should use the normal `is_extensible` function here. `try_is_extensible` is effectively a method that will interrupt its work if it detects that it would need to call into JavaScript. Effectively, that would mean that `target` itself is a Proxy :)

That's probably very unlikely so usually `try_is_extensible` would work fine but it's probably not really necessary here for reasons. This all ties into the interleaved garbage collector that I am slowly working on, but which is not yet turned on by default. But precisely because it's still work-in-progress, you can ignore it here and just use the `is_extensible` function directly which takes a `GcScope` and can thus trigger garbage collection once interleaved GC is turned on. I'll eventually get around to fixing any possible issues that may cause here :)

So in effect, you can just replace this whole if-else with:
```suggestion
        let extensible_target = is_extensible(agent, target, gc.reborrow())?;
```

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+
+        // 10. If extensibleTarget is true, return handlerProto.
+        if extensible_target {
+            return Ok(Object::internal_prototype(
```
**issue:** The spec says "return handlerProto." so `return Ok(handler_proto);` is what you should do, nothing more complicated than that :)

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+        }
+
+        // 11. Let targetProto be ? target.[[GetPrototypeOf]]().
+        let target_proto = Proxy::internal_get_prototype_of(self, agent, gc.reborrow())?;
```
**issue:** `target.[[GetPrototypeOf]]()` converts into Nova as `target.internal_get_prototype_of(agent, gc.reborrow())`. Right now this is a recursive call into self.

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+        let target_proto = Proxy::internal_get_prototype_of(self, agent, gc.reborrow())?;
+
+        // 12. If SameValue(handlerProto, targetProto) is false, throw a TypeError exception.
+        if !same_value(agent, handler_proto, target_proto) {
```
**nitpick:** We happen to know that for object-or-null results like `handler_proto` and `target_proto`, `SameValue` is exactly equal to a direct equality comparison:
```suggestion
        if handler_proto == target_proto {
```

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+        if !same_value(agent, handler_proto, target_proto) {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "Unexpected break in ControlFlow",
```
**nitpick (blocking):** I'm not 100% sure what the error message here should be, but at least it shouldn't be this.

yossydev replied:
Thank you for the review!

Oh no, I seem to have shown you something unsightly. That was left over from when I was tinkering during debugging—apologies for that.

I’ve fixed it now, so could you kindly check again to see if it’s appropriate?
https://github.com/trynova/nova/pull/494/commits/c0feafd4d49916ec8ac0bac5441e3575a2a8737e

aapoalas replied:
It's probably appropriate :)

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+
+        // 13. Return handlerProto.
+        Ok(Object::internal_prototype(
+            Object::try_from(handler_proto).unwrap(),
```
**issue:** Here too, just `Ok(handler_proto)`

aapoalas reviewed (changes requested):
First of all, wow! Internal methods are prickly customers, and Proxy more so than most! Much kudos to picking this up and getting it so well in-hand! <3

I did find some issues though, most of it pretty minor. Once those are fixed, I think this'll be good to merge. I recommend doing the ProxyCreate abstract operation and the Proxy constructor after that, they shouldn't be too hard: You can skip steps 4 and 5 in ProxyCreate due to reasons, and step 3 basically just boils down to a `CreateHeapData` trait's `create` method call using a ProxyHeapData object. That also performs steps 6 and 7 automatically.

aapoalas commented on nova_vm/src/ecmascript/builtins/proxy/abstract_operations.rs:
```diff
@@ -0,0 +1,40 @@
+use crate::{
```
**nitpick:** Add the MPL license header as well.

yossydev replied:
Sorry, I forgot.

aapoalas commented:
Ooops... I got inspired by this, fixed most of my comments, and implemented the Proxy constructor and `[[Get]]` method. I'm sorry.

yossydev commented:
@aapoalas 
Thank you for the review!!!

> Ooops… I got inspired by this, fixed most of my comments, and implemented the Proxy constructor and [[Get]] method. I’m sorry.

Not a problem at all! On the contrary, thank you so much! Having the constructor will help move forward with the test262 tests!

I’ll address the remaining issues you pointed out later. I’ll also continue working on implementing other Proxy functionalities!

yossydev commented on nova_vm/src/ecmascript/builtins/proxy.rs:
```diff
@@ -117,9 +126,95 @@ impl InternalMethods for Proxy {
+        )?;
+
+        // 6. If trap is undefined, then
+        if trap.is_none() {
```
This hadn’t been fixed yet, so I went ahead and corrected it!
https://github.com/trynova/nova/pull/494/commits/aa4cca1d0dd52f0cb4047104b5e02627f51f64e7

aapoalas reviewed (approved):
LGTM, and thank you so much for this! This is great stuff and really puts us forward a lot! <3 Much appreciated
